contracts/libs/data-structures/AvlTree.sol

```diff
@@ -0,0 +1,823 @@
+     */
+    function postOrderTraversal(Bytes32AVL storage tree) internal view returns (bytes32[] memory) {
+        return _getTraversal(tree._tree, _postOrderTraversal);
+    }
+
+    /**
+     * @notice The function to check whether the custom comparator function is set for the bytes32 tree.
+     * @param tree self.
+     * @return True if the custom comparator function is set, false otherwise.
+     */
+    function isCustomComparatorSet(Bytes32AVL storage tree) internal view returns (bool) {
+        return tree._tree.isCustomComparatorSet;
+    }
+
+    /**
+     **********************
+     *     AddressAVL     *
+     **********************
+     */
+
+    struct AddressAVL {
+        Tree _tree;
+    }
+
+    /**
+     * @notice The function to set a custom comparator function, that will be used to build the address tree.
+     * @param tree self.
+     * @param comparator_ The function that accepts keys and values of the nodes to compare.
+     */
+    function setComparator(
+        AddressAVL storage tree,
+        function(bytes32, bytes32, bytes memory, bytes memory) view returns (int8) comparator_
+    ) internal {
+        _setComparator(tree._tree, comparator_);
+    }
+
+    /**
+     * @notice The function to insert a node into the address tree.
+     * Complexity is O(log(n)), where n is the number of elements in the tree.
+     *
+     * @param tree self.
+     * @param key_ The key to insert.
+     * @param value_ The value to insert.
+     */
+    function insert(AddressAVL storage tree, address key_, bytes memory value_) internal {
+        _insert(tree._tree, _asBytes32(key_), value_);
+    }
+
+    /**
+     * @notice The function to remove a node from the address tree.
+     * Complexity is O(log(n)), where n is the number of elements in the tree.
+     *
+     * @param tree self.
+     * @param key_ the key of the node to remove.
+     */
+    function remove(AddressAVL storage tree, address key_) internal {
+        _remove(tree._tree, _asBytes32(key_));
+    }
+
+    /**
+     * @notice The function to search for a node in the address tree.
+     * Complexity is O(1).
+     *
+     * @param tree self.
+     * @param key_ the key of the node to search for.
+     * @return True if the node exists, false otherwise.
+     */
+    function search(AddressAVL storage tree, address key_) internal view returns (bool) {
+        return _search(tree._tree, _asBytes32(key_));
+    }
+
+    /**
+     * @notice The function to retrieve the value associated with a key in the address tree.
+     * Complexity is O(1).
+     *
+     * @param tree self.
+     * @param key_ the key to get the value for.
+     * @return The value associated with the key.
+     */
+    function getValue(
+        AddressAVL storage tree,
+        address key_
+    ) internal view returns (bytes storage) {
+        require(
+            _search(tree._tree, _asBytes32(key_)),
+            "AvlTree: node with such key doesn't exist"
+        );
```

Comment:
Oh nonono. Such reverts are extremely painful for the end users. 

Have two values returned. The actual value and the element's existence (to handle the empty case). The `search` function can be removed then. 

See openzeppelin's enumerable map